firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,301 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                    .transpose();
+
+                Poll::Ready(next)
             }
+        }
+    }
+}
+
+enum ParentNode<'a> {
+    New(ObjRef<'a>),
+    Visited(ObjRef<'a>),
+}
+
+#[derive(Debug)]
+enum InnerNode<'a> {
+    New(&'a NodeType),
+    Visited(&'a NodeType),
+}
+
+impl<'a> ParentNode<'a> {
+    fn inner(&self) -> InnerNode<'_> {
+        match self {
+            Self::New(node) => InnerNode::New(node.inner()),
+            Self::Visited(node) => InnerNode::Visited(node.inner()),
+        }
+    }
+
+    fn into_node(self) -> ObjRef<'a> {
+        match self {
+            Self::New(node) => node,
+            Self::Visited(node) => node,
+        }
+    }
+}
+
+type NextResult = Option<(Vec<u8>, Vec<u8>)>;
+
+fn find_next_result<'a, S: ShaleStore<Node>, T>(
+    merkle: &'a Merkle<S, T>,
+    parents: &mut Vec<(ObjRef<'a>, u8)>,
+) -> Result<NextResult, super::MerkleError> {
+    let next = find_next_node_with_data(merkle, parents)?.map(|(next_node, value)| {
+        let node_path_iter = match next_node.inner() {
+            NodeType::Leaf(leaf) => leaf.path.iter().copied(),
+            NodeType::Extension(extension) => extension.path.iter().copied(),
+            _ => [].iter().copied(),
         };
 
-        // figure out the value to return from the state
-        // if we get here, we're sure to have something to return
-        // TODO: It's possible to return a reference to the data since the last_node is
-        // saved in the iterator
-        let return_value = match &self.key_state {
-            IteratorState::Iterating {
-                last_node,
-                parents: _,
-            } => {
-                let value = match last_node.inner() {
-                    NodeType::Branch(branch) => branch.value.to_owned().unwrap().to_vec(),
-                    NodeType::Leaf(leaf) => leaf.data.to_vec(),
-                    NodeType::Extension(_) => todo!(),
+        let key = key_from_nibble_iter(nibble_iter_from_parents(parents).chain(node_path_iter));
+
+        parents.push((next_node, 0));
+
+        (key, value)
+    });
+
+    Ok(next)
+}
+
+fn find_next_node_with_data<'a, S: ShaleStore<Node>, T>(
+    merkle: &'a Merkle<S, T>,
+    visited_parents: &mut Vec<(ObjRef<'a>, u8)>,
+) -> Result<Option<(ObjRef<'a>, Vec<u8>)>, super::MerkleError> {
+    use InnerNode::*;
+
+    let Some((visited_parent, visited_pos)) = visited_parents.pop() else {
+        return Ok(None);
+    };
+
+    let mut node = ParentNode::Visited(visited_parent);
+    let mut pos = visited_pos;
+    let mut first_loop = true;
+
+    loop {
+        match node.inner() {
+            New(NodeType::Leaf(leaf)) => {
+                let value = leaf.data.to_vec();
+                return Ok(Some((node.into_node(), value)));
+            }
+
+            Visited(NodeType::Leaf(_)) | Visited(NodeType::Extension(_)) => {
+                let Some((next_parent, next_pos)) = visited_parents.pop() else {
+                    return Ok(None);
```

Comment:
I wasn't able to follow the logic here. Can you add some comments as to why we're returning `Ok(None)`?

Reply:
`Fused` is intended (it was easier to write that way now that we aren't calling `std::mem::take` on `IteratorState`. It would be possible to have a cycling stream where we would just wrap the stream with an additional type that resets the state once the end is reached. If we find such a use case, it will be trivial to write. 

As for this particular line you originally commented on, it is interesting. If we assume there is always going to be a sentinel-node, we'll never hit this case. However, if we ignore that assumption, this function doesn't care about the existence of a sentinel node or not. If there are no other nodes to traverse, simply return. It's the more general solution and works in both cases.

Reply:
Oh, and I added the `FusedStream` implementation as suggested.